torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -273,6 +273,34 @@ at::Tensor broadcast(
   return broadcast_(output, src, group_name);
 }
 
+at::Tensor& scatter_(
+    at::Tensor& input,
+    std::vector<at::Tensor> scatter_list,
+    int64_t src,
+    std::string group_name) {
+  c10d::ScatterOptions opts;
+  opts.rootRank = src;
+  std::vector<at::Tensor> outputs{input};
+  std::vector<std::vector<at::Tensor>> inputs{};
+  if (scatter_list.size() > 0) {
+    inputs.push_back(scatter_list);
+  }
+
+  auto group = c10d::resolve_process_group(group_name);
+  auto work = group->scatter(outputs, inputs, opts);
+  c10d::RankLocal<WorkRegistry>::get().register_work(input, work);
+  return input;
+}
+
+at::Tensor scatter(
+    const at::Tensor& input,
+    std::vector<at::Tensor> scatter_list,
+    int64_t src,
+    std::string group_name) {
+  auto output = input.clone(at::MemoryFormat::Contiguous);
```

Comment:
I haven't looked too closely, but I think we don't need to clone the input here right? In all-reduce, we perform the clone because we reuse the input for the output. It's probably not the case for scatter?

Reply:
I do agree that this is likely unnecessary for most usages but I think keeping it for now is best practice to match the other APIs. There is the scatter_ impl that we could expose if we did want to make it a bit more performant. From my previous chats with @wanchaol the "functional" APIs shouldn't have any side effects

Reply:
I'm not saying we should reuse input. I meant, why not `auto output = at::empty_like(input)` here? The copy seems redundant in any scenario.

Reply:
ahhh, you're totally right -- done